src/sparse_int_set.jl

```diff
@@ -0,0 +1,258 @@
+		s.packed[packed_id]               = packed_endid
+		s.reverse[to_page][to_offset]     = s.reverse[from_page][from_offset]
+		s.reverse[from_page][from_offset] = 0
+		pop!(s.packed)
+	end
+    return id
+end
+
+function pop!(s::SparseIntSet, id::Integer, default)
+	id < 0 && throw(ArgumentError("Int to pop needs to be positive."))
+	in(id, s) ? (@inbounds pop!(s, id)) : default
+end
+popfirst!(s::SparseIntSet) = pop!(s, first(s))
+
+iterate(set::SparseIntSet, args...) = iterate(set.packed, args...) 
+
+function empty!(s::SparseIntSet)
+	empty!(s.packed)
+	for p in s.reverse
+		fill!(p, 0)
+	end
+	return s
+end
+
+last(s::SparseIntSet) = isempty(s) ? throw(ArgumentError("Empty set has no last element.")) : last(s.packed)
+
+union(s::SparseIntSet, ns) = union!(copy(s), ns)
+union!(s::SparseIntSet, ns) = (for n in ns; push!(s, n); end; s)
+
+intersect(s1::SparseIntSet) = copy(s1)
+intersect(s1::SparseIntSet, ss...) = intersect(s1, intersect(ss...))
+function intersect(s1::SparseIntSet, ns)
+	s = SparseIntSet()
+    for n in ns
+        n in s1 && push!(s, n)
+    end
+    return s
+end
+
+intersect!(s1::SparseIntSet, ss...) = intersect!(s1, intersect(ss...))
+
+#Is there a more performant way to do this?
+function intersect!(s1::SparseIntSet, ns)
+	s = intersect(s1, ns)
+	s1.packed= s.packed
+	s1.reverse = s.reverse
+	return s1
+end
+
+setdiff(s::SparseIntSet, ns) = setdiff!(copy(s), ns)
+setdiff!(s::SparseIntSet, ns) = (for n in ns; pop!(s, n, nothing); end; s)
+
+function ==(s1::SparseIntSet, s2::SparseIntSet)
+	length(s1) != length(s2) && return false
+	return all(x -> in(x, s1), s2)
+end
+
+issubset(a::SparseIntSet, b::SparseIntSet) = isequal(a, intersect(a, b))
+
+complement(a::SparseIntSet) = complement!(SparseIntSet(), a)
+function complement!(b::SparseIntSet, a::SparseIntSet)
+	empty!(b)
+	for i in eachindex(a.reverse)
+		if !isassigned(a.reverse, i)
+			resize!(b.reverse, i)
+			new_ids = (i-1)*INT_PER_PAGE+1:(i)*INT_PER_PAGE
+			append!(b.packed, new_ids)
+			b.reverse[i] = collect(new_ids)
+		else
+			for offset in 1:INT_PER_PAGE
+				if a.reverse[i][offset] == 0
+					push!(b, INT_PER_PAGE*(i-1) + offset)
+				end
+			end
+		end
+	end
+	return b
+end
+#Can this be optimized?
+complement!(a::SparseIntSet) = (b = complement(a); a.packed = b.packed; a.reverse = b.reverse; return a)
+
+<(a::SparseIntSet, b::SparseIntSet) = (a<=b) && !isequal(a,b)
+<=(a::SparseIntSet, b::SparseIntSet) = issubset(a, b)
+
+function findfirst_packed_id(i, s::SparseIntSet)
+	pageid, offset = pageid_offset(s, i)
+	if isassigned(s.reverse, pageid)
+		@inbounds id = s.reverse[pageid][offset]
+		return id
+	end
+	return 0
+end
+
+function cleanup!(s::SparseIntSet)
+	isused = x -> isassigned(s.reverse, x) && any(y -> y != 0, s.reverse[x])
+	indices = eachindex(s.reverse)
+	last_page_id = findlast(isused, indices)
+	if last_page_id === nothing
+		s.reverse = Vector{Int}[]
+		return s
+	else
+		new_pages    = Vector{Vector{Int}}(undef, last_page_id)
```

Comment:
Spacing is extra weird here

Reply:
Hmm right, I think somehow kakoune is using tabs for indentation. This never seemed to have been an issue before but now somehow it does... I'll investigate and change.